packages/gatsby/cache-dir/__tests__/dev-loader-lazy.js

```diff
@@ -0,0 +1,550 @@
+    })
+
+    it(`returns page resources when already fetched`, () => {
+      const devLoader = new DevLoader(null, [])
+
+      expect(devLoader.loadPageSync(`/mypage/`)).toBeUndefined()
+    })
+  })
+
+  describe(`prefetch`, () => {
+    const flushPromises = () => new Promise(resolve => setImmediate(resolve))
+
+    it(`shouldn't prefetch when shouldPrefetch is false`, () => {
+      const devLoader = new DevLoader(null, [])
+      devLoader.shouldPrefetch = jest.fn(() => false)
+      devLoader.doPrefetch = jest.fn()
+      devLoader.apiRunner = jest.fn()
+
+      expect(devLoader.prefetch(`/mypath/`)).toBe(false)
+      expect(devLoader.shouldPrefetch).toHaveBeenCalledWith(`/mypath/`)
+      expect(devLoader.apiRunner).not.toHaveBeenCalled()
+      expect(devLoader.doPrefetch).not.toHaveBeenCalled()
+    })
+
+    it(`should trigger custom prefetch logic when core is disabled`, () => {
+      const devLoader = new DevLoader(null, [])
+      devLoader.shouldPrefetch = jest.fn(() => true)
+      devLoader.doPrefetch = jest.fn()
+      devLoader.apiRunner = jest.fn()
+      devLoader.prefetchDisabled = true
+
+      expect(devLoader.prefetch(`/mypath/`)).toBe(false)
+      expect(devLoader.shouldPrefetch).toHaveBeenCalledWith(`/mypath/`)
+      expect(devLoader.apiRunner).toHaveBeenCalledWith(`onPrefetchPathname`, {
+        pathname: `/mypath/`,
+      })
+      expect(devLoader.doPrefetch).not.toHaveBeenCalled()
+    })
+
+    it(`should prefetch when not yet triggered`, async () => {
+      jest.useFakeTimers()
+      const devLoader = new DevLoader(null, [])
+      devLoader.shouldPrefetch = jest.fn(() => true)
+      devLoader.apiRunner = jest.fn()
+      devLoader.doPrefetch = jest.fn(() => Promise.resolve({}))
+
+      expect(devLoader.prefetch(`/mypath/`)).toBe(true)
+
+      // wait for doPrefetchPromise
+      await flushPromises()
+
+      expect(devLoader.apiRunner).toHaveBeenCalledWith(`onPrefetchPathname`, {
+        pathname: `/mypath/`,
+      })
+      expect(devLoader.apiRunner).toHaveBeenNthCalledWith(
+        2,
+        `onPostPrefetchPathname`,
+        {
+          pathname: `/mypath/`,
+        }
+      )
+    })
+
+    it(`should only run apis once`, async () => {
+      const devLoader = new DevLoader(null, [])
+      devLoader.shouldPrefetch = jest.fn(() => true)
+      devLoader.apiRunner = jest.fn()
+      devLoader.doPrefetch = jest.fn(() => Promise.resolve({}))
+
+      expect(devLoader.prefetch(`/mypath/`)).toBe(true)
+      expect(devLoader.prefetch(`/mypath/`)).toBe(true)
+
+      // wait for doPrefetchPromise
+      await flushPromises()
+
+      expect(devLoader.apiRunner).toHaveBeenCalledTimes(2)
+      expect(devLoader.apiRunner).toHaveBeenNthCalledWith(
+        1,
+        `onPrefetchPathname`,
+        expect.anything()
+      )
+      expect(devLoader.apiRunner).toHaveBeenNthCalledWith(
+        2,
+        `onPostPrefetchPathname`,
+        expect.anything()
+      )
+    })
+  })
+})
```

Comment:
Just comparing those tests to current `dev-loader.js` it seems like it has only minor adjustments for setup (setting `GATSBY_EXPERIMENT_LAZY_DEVJS` env var).

Changing the field name in `createSyncRequires` helper and ... removes one test:

```
    it(`should return an error when component cannot be loaded`, async () => {
      const syncRequires = createSyncRequires({
        chunk: false,
      })
      const devLoader = new DevLoader(syncRequires, [])
      const pageData = {
        path: `/mypage/`,
        componentChunkName: `chunk`,
        staticQueryHashes: [],
      }
      devLoader.loadPageDataJson = jest.fn(() =>
        Promise.resolve({
          payload: pageData,
          status: `success`,
        })
      )

      await devLoader.loadPage(`/mypage/`)
      const expectation = devLoader.pageDb.get(`/mypage`)
      expect(expectation).toHaveProperty(`status`, `error`)
      expect(emitter.emit).toHaveBeenCalledTimes(0)
    })
```

I'm not sure of the value of that test to begin with but it looks a bit look a smell that something doesn't behave the same?

Reply:
This is an intended change — w/ lazy bundling — it isn't easy to know if a page component can't be loaded. Perhaps we could add a timeout or something but it didn't seem like this made sense to test against.